### ISSUE TITLE
Improved torch.func support in Opacus

### ISSUE THREAD

**Issue by eddiestudies** [meta-pytorch/opacus#846]
## 🚀 Feature
Adding a "func" mode for Opacus
## Motivation

per-sample gradients for models that fit in memory reduce the proof surface area. In an attempt to better understand what is possible, I evaluated torch.func for many model types. I then tried to mirror jax-privacy's approach to see if it was compatible with opacus.

The goal is to have a faster default that is simpler to maintain by not requiring per layer changes in Opacus. 
## Pitch

# Simplest approach, requires a loss_fn in vmap form
```python
assert False, "This code fails, provided for feedback"
def loss_fn(model, x, y):
    return F.cross_entropy(model(x), y)

model, optimizer, dp_loss, loader = engine.make_private(
    module=model, 
    optimizer=optimizer, 
    data_loader=loader, 
    loss_fn=loss_fn,
    noise_multiplier=1.0, 
    max_grad_norm=1.0, 
    grad_sample_mode="func_hybrid") # Would call it func

for x, y in loader:
    optimizer.zero_grad()
    loss = dp_loss(x, y)               # holds the batch; lazy, but in a more expected way
    loss.backward()                     # vmap(grad) + clip + sum -> p.grad
    optimizer.step()                     # noise + accountant, as for ghost
```

## Alternatives

```python
# mirror of ghost but it requires a lazy loss output that felt off to me, it mirrored the api calls
assert False, "This code fails, provided for feedback"
model, optimizer, criterion, loader = engine.make_private(
    module=model, optimizer=optimizer, data_loader=loader,
    criterion=nn.CrossEntropyLoss(),
    noise_multiplier=1.0, max_grad_norm=1.0, grad_sample_mode="func")

for x, y in loader:
    optimizer.zero_grad()
    loss = criterion(model(x), y)      # model(x) is a LazyDPOutput: no forward ran
    loss.backward()                          # the fold runs here
    optimizer.step()

```

```python
# This was an attempt to see if we could mirror jax-privacy more closely
# This approach is not idiomatic to opacus but seemed worth understanding
clipped = opacus.clipped_grad(model, loss_fn, max_grad_norm=1.0, chunk_size=16)
accountant = RDPAccountant()

for x, y in loader:
    grad_sum = clipped(x, y)                      # {name: clipped sum}
    for n, p in model.named_parameters():
        p.grad = (grad_sum[n] + σ*C*torch.randn_like(p)) / B
    optimizer.step()                              # a plain torch optimizer
    accountant.step(noise_multiplier=σ, sample_rate=q)
```

## Additional context
I am willing to contribute this change. I have been locally testing the idea across model zoos for a few weeks. It is an attempt to improve the privacy posture of Opacus by improving per-sample gradient handling.

There are a few torch related issues that can further improve support. I found an in place operation in 2 Conv* layers that I am looking to upstream to PyTorch.
There are two nlp related graph breaks that hurt performance. Both are solvable with Aten related PRs, effectively teaching torch.func how to handle the batching for vmap. Both will be upstreamed as well.
This avoids a  Hugging Face integration nuance I found where certain HuggingFace layers, LayerNorm for example, use a different class that circumvents the built in handling in Opacus.

**Comment by eddiestudies** [comment]
I created a notebook that isolates the torch.func code at the top:

https://colab.research.google.com/drive/1bsU2_wY2VXJ5umav8Ej-Wrzs4XguDuQD#scrollTo=p9c-eUef4iaI

I have been validating the change for a few weeks now locally across many models. The performance and memory improvements over hooks has been very consistent. ghost is better at larger model sizes but, and the change is not compatible with FSDP yet. 

The idea here is that torch.func might solve most, if not all, smaller model scenarios with no performance hit over other options. While ghost becomes the solution for larger model sizes that do not fit in memory.

I looked into BK, however, this torch.func approach seemed to be faster and more memory efficient up until ghost was the right solution. I did not expect that after reading the literature but it was a nice surprise.

**Comment by eddiestudies** [comment]
@iden-kalemaj  this is a change I was looking to contribute. The idea is to find a faster torch.func based solution in Opacus. There have been many improvements to the torch.func API and I found it to be faster and more memory efficient than most existing solutions in Opacus.

My main questions are around desired API. I tried to mirror the ghost API closely and having a lazy output to exactly mirror it seemed a bit off. There is a somewhat related approach in Opacus but it seemed as if I was overdoing it to match the exact ghost API. 

I tried evaluating the chanege across many models and it seemed to work well but I do not have full context on model coverage. Most layers were supported out of the box except for BatchNorm. Which made sense. A few computer vision layers, similar to the ones you patched with the in-place fix also failed. 

The in-place fix prompted my looking into this more closely. It seemed possible for custom user code to interact poorly with in-place handling in ModuleValidator since custom user code avoids a lot of the Opacus machinery. torch.func also has trouble with in-place operations, however most models do not have them for the torch.func path. Two computer vision layers have an inplace op, I am opening a PR to try to patch those in parallel to these changes.

tldr; I have a torch.func change that is faster than most in memory options in Opacus, it might require a slight API change or a lazy criterion implementation to mirror ghost's API exactly. 

Is it better to decide on the API in this issue or for me to make a PR with the options and then through review we land at the right one if one exists?
